### PR TITLE
Revise pyro to match pytorch master

### PR DESCRIPTION
**Pull request by fehiepsi**
This pull request revises the following points
+ Set default values for `loc` and `scale` of HalfCauchy, so when updating, it is enough to delete the implementation of HalfCauchy.
+ Add `dim` argument for `log_sum_exp` operator, so when updating, it is enough to change its name to `torch.logsumexp`
+ Make `matrix_triagular_solve_compat` function hidden inside `lowrank_mvn`. So when updating, it is enough to delete the implementation of lowrank_mvn. 
+ Use `torch.trtrs` directly in GP module
+ Use `.new_tensor`, `.new_empty` instead of `.new` at various places
+ Make the sum of a matrix with a diagonal matrix faster by using `view` trick.

There is still one thing needed to be updated relevant to `trace_term` of lowrank_mvn. I will address it in the other pull request.

### CODE REVIEW

File: tests/contrib/gp/test_models.py [hunk 1]
```diff
@@ -330,9 +330,10 @@ def _pre_test_mean_function():
     def f(x):
         return 2 * x + 3 + 5 * torch.sin(7 * x)
 
-    X = torch.arange(100)
+    tensor_holder = torch.tensor([])
```
Comment (fritzo) [inline tests/contrib/gp/test_models.py]: Can you explain the change in this file? I don't understand why it's needed.

Comment (fehiepsi) [reply tests/contrib/gp/test_models.py]: I used pytorch master, and `torch.arange` returns long tensor instead of float as in pytorch 0.4.

**Comment by eb8680** [comment]
@fehiepsi did you actually run any Pyro tests against PyTorch master?  I'm seeing a bunch of our tests in `tests/infer/test_valid_models.py` and `tests/infer/test_enum.py` failing due to deprecation warnings.

**Comment by fehiepsi** [comment]
@eb8680 I just run the gp model tests and observe some deprecation warning about bce_with_logits. So I made a pull request in pytorch regarding that warning. I didn't run other Pyro tests.

I will run the tests in those two files to update more.

**Comment by eb8680** [comment]
I'm not sure the failures have anything to do with this PR, but we probably don't want Pyro dev to be tracking PyTorch master too closely because of the additional maintenance burden.  I guess it's OK for `contrib` modules, though, and the changes in this PR and #1212 seem reasonable.

Also, FYI, those two test modules I mentioned are very useful for development - they have a lot of small, reliable tests that run very quickly and cover a large subset of core SVI and broadcasting functionality.  I run them all the time when I'm making changes to core Pyro to catch bugs I've introduced.

**Comment by fehiepsi** [comment]
@eb8680 I run the test with pytorch master and do not observe the deprecate warnings for those two files. Maybe the warnings you got were related to Bernoulli distribution, which I have made a fix (which has been merged) in pytorch master.
